File: python/mirage/core/linear/pathing.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.utils.naming import SEPARATOR, fit_id_name, parse_id_name
from mirage.utils.sanitize import sanitize_name

split_suffix_id = parse_id_name


def team_dirname(team: dict[str, Any]) -> str:
    parts: list[str] = []
    if team.get("key"):
        parts.append(sanitize_name(team["key"]))
    if team.get("name"):
        sanitized_name = sanitize_name(team["name"])
        if sanitized_name not in parts:
            parts.append(sanitized_name)
    if not parts:
        parts.append("team")
    # fit_id_name rather than make_id_name: the parts are already
    # sanitized and joined with the separator, and re-sanitizing would
    # collapse that `__` to `_`.
    return fit_id_name(SEPARATOR.join(parts), team["id"])


def member_filename(user: dict[str, Any]) -> str:
    label = sanitize_name(
        user.get("displayName") or user.get("name") or user.get("email")
        or "user")
    return fit_id_name(label, user["id"], ".json")


def issue_dirname(issue: dict[str, Any]) -> str:
    key = issue.get("identifier") or issue.get("id") or "issue"
    return fit_id_name(sanitize_name(key), issue["id"])


def project_filename(project: dict[str, Any]) -> str:
    label = sanitize_name(project.get("name") or "project")
    return fit_id_name(label, project["id"], ".json")


def cycle_filename(cycle: dict[str, Any]) -> str:
    label = sanitize_name(cycle.get("name") or "cycle")
    return fit_id_name(label, cycle["id"], ".json")


def document_filename(document: dict[str, Any]) -> str:
    label = sanitize_name(document.get("title") or "document")
    return fit_id_name(label, document["id"], ".json")

File: typescript/packages/core/src/core/linear/pathing.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { sanitizeName } from '../../utils/sanitize.ts'

export { sanitizeName } from '../../utils/sanitize.ts'
export { parseIdName as splitSuffixId } from '../../utils/naming.ts'
import { SEPARATOR, fitIdName } from '../../utils/naming.ts'

function pickString(record: Record<string, unknown>, ...keys: readonly string[]): string {
  for (const key of keys) {
    const value = record[key]
    if (typeof value === 'string' && value !== '') return value
  }
  return ''
}

function requireId(record: Record<string, unknown>): string {
  const id = pickString(record, 'id')
  if (id === '') throw new Error('record missing id')
  return id
}

export function teamDirname(team: Record<string, unknown>): string {
  const parts: string[] = []
  const key = pickString(team, 'key')
  if (key !== '') parts.push(sanitizeName(key))
  const name = pickString(team, 'name')
  if (name !== '') {
    const sanitized = sanitizeName(name)
    if (!parts.includes(sanitized)) parts.push(sanitized)
  }
  if (parts.length === 0) parts.push('team')
  // fitIdName rather than makeIdName: the parts are already sanitized and
  // joined with the separator, and re-sanitizing would collapse that `__`
  // to `_`.
  return fitIdName(parts.join(SEPARATOR), requireId(team))
}

export function memberFilename(user: Record<string, unknown>): string {
  const label = sanitizeName(pickString(user, 'displayName', 'name', 'email') || 'user')
  return fitIdName(label, requireId(user), '.json')
}

export function issueDirname(issue: Record<string, unknown>): string {
  const key = pickString(issue, 'identifier', 'id') || 'issue'
  return fitIdName(sanitizeName(key), requireId(issue))
}

export function projectFilename(project: Record<string, unknown>): string {
  const label = sanitizeName(pickString(project, 'name') || 'project')
  return fitIdName(label, requireId(project), '.json')
}

export function cycleFilename(cycle: Record<string, unknown>): string {
  const label = sanitizeName(pickString(cycle, 'name') || 'cycle')
  return fitIdName(label, requireId(cycle), '.json')
}

export function documentFilename(document: Record<string, unknown>): string {
  const label = sanitizeName(pickString(document, 'title') || 'document')
  return fitIdName(label, requireId(document), '.json')
}
